internal/tests/api/targets/target_test.go

```diff
@@ -379,6 +380,166 @@ func TestList(t *testing.T) {
+	require.NoError(t, err)
+	require.NotNil(t, hUpdate)
+
+	// Expect error when associating a host source to the target with a direct network address association
+	targetId := targetResp.GetItem().Id
+	version := targetResp.GetItem().Version
+	updateResp, err := tClient.AddHostSources(tc.Context(), targetId, version, []string{hs.Item.Id})
+	require.Error(t, err)
+	require.Nil(t, updateResp)
+
+	// Remove direct network address association. Successfully add a host source to target.
+	targetResp, err = tClient.Update(tc.Context(), targetId, version, targets.DefaultAddress())
+	require.NoError(t, err)
+	require.NotNil(t, targetResp)
+	require.Empty(t, targetResp.GetItem().Address)
+	version = targetResp.GetItem().Version
+	updateResp, err = tClient.AddHostSources(tc.Context(), targetId, version, []string{hs.Item.Id})
+	require.NoError(t, err)
+	require.NotNil(t, updateResp)
+	require.Empty(t, updateResp.GetItem().Address)
+	require.Equal(t, []string{hs.Item.Id}, updateResp.GetItem().HostSourceIds)
+}
+
+func TestTarget_HostSourceMutualExlcusiveRelationShip(t *testing.T) {
+	tc := controller.NewTestController(t, nil)
+	t.Cleanup(tc.Shutdown)
+
+	client := tc.Client()
+	at := tc.Token()
+	client.SetToken(at.Token)
+	_, proj := iam.TestScopes(t, tc.IamRepo(), iam.WithUserId(at.UserId))
+	tClient := targets.NewClient(client)
+
+	// Setup host catalog, host set, & host resources
+	hc, err := hostcatalogs.NewClient(client).Create(tc.Context(), "static", proj.GetPublicId())
+	require.NoError(t, err)
+	require.NotNil(t, hc)
+	hs, err := hostsets.NewClient(client).Create(tc.Context(), hc.Item.Id)
+	require.NoError(t, err)
+	require.NotNil(t, hs)
+	h, err := hosts.NewClient(client).Create(tc.Context(), hc.Item.Id, hosts.WithStaticHostAddress("localhost"))
+	require.NoError(t, err)
+	require.NotNil(t, h)
+	hUpdate, err := hostsets.NewClient(client).AddHosts(tc.Context(), hs.Item.Id, hs.Item.Version, []string{h.GetItem().Id})
+	require.NoError(t, err)
+	require.NotNil(t, hUpdate)
+
+	// Create target without a network address association
+	targetResp, err := tClient.Create(tc.Context(), "tcp", proj.GetPublicId(),
+		targets.WithName("test-host-source"), targets.WithTcpTargetDefaultPort(22))
+	require.NoError(t, err)
+	require.NotNil(t, targetResp)
+	require.Empty(t, targetResp.GetItem().Address)
+
+	// Expect error when associating a address to the target with a host source association
+	targetId := targetResp.GetItem().Id
+	version := targetResp.GetItem().Version
+	updateResp, err := tClient.AddHostSources(tc.Context(), targetId, version, []string{hs.Item.Id})
+	require.NoError(t, err)
+	require.NotNil(t, updateResp)
+	require.Empty(t, updateResp.GetItem().Address)
+	require.Equal(t, []string{hs.Item.Id}, updateResp.GetItem().HostSourceIds)
+	version = updateResp.GetItem().Version
+	updateResp, err = tClient.Update(tc.Context(), targetId, version, targets.WithAddress("localhost"))
+	require.Error(t, err)
```

Comment:
Ditto, check the error type

Reply:
Addressed [40dfc1f](https://github.com/hashicorp/boundary/pull/2613/commits/40dfc1f9f3a0d16b1934d67583b89a695d7318b7)